usersession/agent/session_agent.go

```diff
@@ -79,6 +84,63 @@ func (c *Command) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	rsp.ServeHTTP(w, r)
 }
 
+type idleTracker struct {
+	mu     sync.Mutex
+	active map[net.Conn]struct{}
+	change chan struct{}
+}
+
+var sysGetsockoptUcred = sys.GetsockoptUcred
+
+func getUcred(conn net.Conn) (*sys.Ucred, error) {
+	if uconn, ok := conn.(*net.UnixConn); ok {
+		f, err := uconn.File()
+		if err != nil {
+			return nil, err
+		}
+		defer f.Close()
+		return sysGetsockoptUcred(int(f.Fd()), sys.SOL_SOCKET, sys.SO_PEERCRED)
+	}
+	return nil, fmt.Errorf("cannot get peer credentials for connection %v", conn)
+}
+
+func (it *idleTracker) trackConn(conn net.Conn, state http.ConnState) {
+	// Perform peer credentials check
+	if state == http.StateNew {
+		ucred, err := getUcred(conn)
+		uid := uint32(sys.Geteuid())
+		if err != nil || (ucred.Uid != 0 && ucred.Uid != uid) {
+			conn.Close()
+		}
+		return
+	}
+
+	it.mu.Lock()
+	oldActive := len(it.active)
+	if state == http.StateNew || state == http.StateActive {
+		it.active[conn] = struct{}{}
+	} else {
+		delete(it.active, conn)
+	}
+	newActive := len(it.active)
+	it.mu.Unlock()
```

Comment:
The aim was to release the lock before sending on the channel to avoid a deadlock.  But thinking about things more, I think things are still open to deadlocking during shutdown: if `exitOnIdle()` kills the tomb, it won't be reading from the `idleTracker` channel.  That could cause a deadlock if a connection came in between then and when `http.Server.Shutdown()` is called.  I'll have to think about this a bit more.

Reply:
Thanks for this comment! Maybe we can add a small comment in the code like: `// release the lock before sending to the channel to avoid deadlock` or similar (or whatever is appropriate after you rethought this bit :)

Reply:
Okay, so I've changed the way things work now:

1. `idleTracker` now has a `lastActive` field holding a timestamp that is updated each time the number of active connections changes to 0.
2. a new `idleTracker.idleDuration` method returns the length of time server has been idle, or 0 if there are active connections.
3. `exitOnIdle` now only selects on the tomb and the timer.  When the timer expires, we exit if the idle duration is greater than or equal to the timeout.  If it is less, we reset the timer appropriately (e.g. if the timeout is 30 seconds and the server has been idle for 5 seconds, we reset the timer for 25 seconds).

This gets rid of the synchronisation problem, and probably results in fewer wakeups overall.